gossip/src/crds_value.rs

```diff
@@ -490,84 +496,193 @@ impl Sanitize for NodeInstance {
+            .zip([1, 0].iter().cycle())
+            .flat_map(|(bit_count, bit)| std::iter::repeat(bit).take(bit_count.0.into()))
+            .enumerate()
+            .filter_map(|(offset, bit): (usize, &i32)| {
+                if *bit > 0 {
+                    Some(last_slot.saturating_sub(offset as Slot))
+                } else {
+                    None
+                }
+            })
+            .take(MAX_SLOTS_PER_ENTRY)
+            .take_while(|slot| *slot >= min_slot)
+            .collect();
+        result.reverse();
+        result
+    }
+}
+
+#[derive(Deserialize, Serialize, Clone, Debug, PartialEq, Eq, AbiExample)]
+pub struct RawOffsets {
+    bits: BitVec<u8>,
+}
+
+impl RawOffsets {
+    pub fn new(mut bits: BitVec<u8>) -> Self {
+        bits.truncate(RestartLastVotedForkSlots::MAX_SPACE as u64 * 8);
+        bits.shrink_to_fit();
+        Self { bits }
+    }
+
+    pub fn to_slots(&self, last_slot: Slot, min_slot: Slot) -> Vec<Slot> {
+        let mut result = Vec::new();
+        for i in 0..self.bits.len() {
+            if self.bits.get(i) {
+                let slot = last_slot.saturating_sub(i as Slot);
+                if slot < min_slot {
+                    return result;
+                }
+                result.push(slot);
+                if result.len() > MAX_SLOTS_PER_ENTRY {
+                    break;
+                }
+            }
+        }
+        result.reverse();
+        result
+    }
+}
+
+#[derive(Serialize, Deserialize, Clone, PartialEq, Eq, AbiExample, Debug)]
 pub struct RestartLastVotedForkSlots {
     pub from: Pubkey,
     pub wallclock: u64,
-    pub slots: Vec<CompressedSlots>,
+    offsets: SlotsOffsets,
+    pub last_voted_slot: Slot,
     pub last_voted_hash: Hash,
     pub shred_version: u16,
 }
 
 impl Sanitize for RestartLastVotedForkSlots {
     fn sanitize(&self) -> std::result::Result<(), SanitizeError> {
-        if self.slots.is_empty() {
-            return Err(SanitizeError::InvalidValue);
-        }
-        self.slots.sanitize()?;
         self.last_voted_hash.sanitize()
     }
 }
 
 impl RestartLastVotedForkSlots {
-    pub fn new(from: Pubkey, now: u64, last_voted_hash: Hash, shred_version: u16) -> Self {
-        Self {
+    // This number is MAX_CRDS_OBJECT_SIZE - empty serialized RestartLastVotedForkSlots.
+    const MAX_SPACE: usize = 824;
+
+    pub fn new(
+        from: Pubkey,
+        now: u64,
+        last_voted_fork: &[Slot],
+        last_voted_hash: Hash,
+        shred_version: u16,
+    ) -> Result<Self, String> {
+        let (first_voted_slot, last_voted_slot) = match last_voted_fork.iter().minmax() {
+            NoElements => return Err("Last voted slot must be specified".to_string()),
+            OneElement(slot) => (slot, slot),
+            MinMax(min, max) => (min, max),
+        };
+        let max_size = last_voted_slot.saturating_sub(*first_voted_slot) + 1;
+        let mut uncompressed_bitvec = BitVec::new_fill(false, max_size);
```

Comment:
`BitVec::new` instead of `new_fill`:
https://docs.rs/bv/latest/bv/struct.BitVec.html#method.new

Reply:
Hmm, this doesn't work, set_bit all panicked with out of bounds. I think new and with_capacity all keep len at 0 so set will fail.